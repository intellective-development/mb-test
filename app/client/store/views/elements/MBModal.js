// @flow

import * as React from 'react';
import OverlayModal from 'react-overlays/lib/Modal';
import bindClassNames from '../../../shared/utils/bind_classnames';
import MBIcon from './MBIcon';
import * as MBText from './MBText';
import styles from './MBModal.scss';

const cn = bindClassNames(styles);

// This module exports a reactive modal and a subcomponents.
// The idea is to provide a new baseline for us to work off of as we move towards component style sharing.
// Eventually, the goal is to fit this component and its props into the styleguide the design team is working towards.

const MODAL_SIZE_CLASSES = {
  tiny: 'elMBModal_Content__Tiny',
  small: 'elMBModal_Content__Small',
  medium: 'elMBModal_Content__Medium',
  large: 'elMBModal_Content__Large'
};

type MBModalProps = {
  children: React.Node,
  onHide?: Function,
  show: boolean,
  size: 'tiny' | 'small' | 'medium' | 'large',
  modalClassName?: string
}
export class Modal extends React.Component<MBModalProps> {
  static defaultProps = {size: 'large'};

  preventHide = (event: SyntheticInputEvent<*>) => {
    event.stopPropagation();
  }

  render(){
    const { size, modalClassName, show, onHide, children } = this.props;
    const classes = cn('elMBModal_Content', MODAL_SIZE_CLASSES[size], modalClassName);

    return (
      <OverlayModal
        backdropClassName={cn('elMBModal_Backdrop')}
        show={show}
        onHide={onHide} >
        <div className={cn('elMBModal_Wrapper')} onClick={onHide}>
          <div className={classes} onClick={this.preventHide}>
            {children}
          </div>
        </div>
      </OverlayModal>
    );
  }
}

type SectionHeaderProps = {className?: string, top?: boolean, children?: React.Node, renderLeft?: () => React.Node, renderRight?: () => React.Node};
export const SectionHeader = ({className, top = true, children, renderLeft = SectionHeaderFiller, renderRight = SectionHeaderFiller}: SectionHeaderProps) => {
  const classes = cn('elMBModal_SectionHeader', className, {
    elMBModal_SectionHeader_Top: top
  });

  return (
    <div className={classes}>
      {renderLeft()}
      <MBText.H4 className={cn('elMBModal_SectionHeaderText')}>{children}</MBText.H4>
      {renderRight()}
    </div>
  );
};

const SectionHeaderFiller = () => <div className={cn('elMBModal_SectionHeader_Filler')} />;

type CloseProps = {onClick?: Function};
export const Close = ({onClick}: CloseProps) => (
  <div className={cn('elMBModal_Close')} onClick={onClick} aria-label="Close">
    <MBIcon name="close" />
  </div>
);

type BackProps = {onClick?: Function};
export const Back = ({onClick}: BackProps) => (
  <div className={cn('elMBModal_Back')} onClick={onClick} aria-label="Back">
    <MBIcon name="back" />
    <MBText.Span className={cn('elMBModal_BackText')}>&nbsp;Back</MBText.Span>
  </div>
);
