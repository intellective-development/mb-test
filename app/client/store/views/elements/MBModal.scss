@import "../style/variables";
@import "../../../../assets/stylesheets/minibar/animations/_splat";

@mixin fullpage-modal-component {
  position: fixed;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
}

.elMBModal_Backdrop {
  @include fullpage-modal-component;

  background-color: $brand-black;
  opacity: 0.5;
  z-index: 10040;
}

.elMBModal_Wrapper {
  @include fullpage-modal-component;

  // these overflow rules allow vertical scrolling
  overflow-x: hidden;
  overflow-y: auto;
  z-index: 10041;
}

$modal-radius: 6px;
// on some resolutions, the combination of border-radius, overflow,
// and margin props is adding a 1px border on the right
.elMBModal_Content {
  @include fullpage-modal-component;

  animation: animationSplat 850ms linear both;
  background-color: white;
  border-radius: $modal-radius;
  box-shadow: 0 2px 4px 0 rgba(0, 0, 0, 0.5);
  margin: 100px auto 32px;
  max-width: $grid-breakpoint;
  padding: 0;
  position: relative;
  width: 70%;
  &:focus { outline: none; }
}
.elMBModal_Content__Tiny { width: 30%; }
.elMBModal_Content__Small { width: 40%; }
.elMBModal_Content__Medium { width: 60%; }
.elMBModal_Content__Large { width: 70%; }

@media screen and (max-width: $small-max-width) {
  .elMBModal_Content { width: 100%; }
  .elMBModal_Content__Tiny { width: 60%; }
  .elMBModal_Content__Small { width: 80%; }

  .elMBModal_Content__Medium,
  .elMBModal_Content__Large {
    border-radius: 0;
    width: 100%;
  }
}

.elMBModal_SectionHeader {
  align-items: center;
  background-color: $brand-light-grey;
  display: flex;
  padding: 6px $standard-gutter;
}

.elMBModal_SectionHeader_Top {
  border-top-left-radius: $modal-radius;
  border-top-right-radius: $modal-radius;
}

.elMBModal_SectionHeaderText {
  @include font-size(14);

  color: $brand-dark-grey;
  flex: auto;
  margin-top: 2px; // visually center it
  text-align: center;
  // text-transform: uppercase; // drop uppercase as requested by Lauren from Minibar
}

$section-side-width: 60px;

.elMBModal_SectionHeader_Filler {
  width: $section-side-width;
}

.elMBModal_Close {
  cursor: pointer;
  display: flex;
  justify-content: flex-end;
  width: $section-side-width;

  &:hover,
  &:focus {
    opacity: 0.6;
  }
}

.elMBModal_Back {
  align-items: center;
  cursor: pointer;
  display: flex;
  width: $section-side-width;

  &:hover,
  &:focus {
    opacity: 0.6;
  }
}

.elMBModal_BackText {
  @include font-size(14);

  font-weight: $heavy-font-weight;
  margin-top: 2px; // visually center it
  text-transform: uppercase;
}
