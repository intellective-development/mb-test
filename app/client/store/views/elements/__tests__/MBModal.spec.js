
import * as React from 'react';

// import TestProvider from 'client/store/__tests__/utils/TestProvider';
import * as MBModal from '../MBModal';

describe('MBModal.Modal', () => {
  it('renders', () => {
    expect(shallow(
      <MBModal.Modal >
        <div>My Content!</div>
      </MBModal.Modal>
    )).toMatchSnapshot();
  });

  it('renders with a header in its content', () => {
    expect(shallow(
      <MBModal.Modal >
        <MBModal.SectionHeader>
          Look at this modal
        </MBModal.SectionHeader>
        <div>My Content!</div>
      </MBModal.Modal>
    )).toMatchSnapshot();
  });
});

describe('MBModal.SectionHeader', () => {
  it('renders', () => {
    expect(shallow(
      <MBModal.SectionHeader>
        Look at this modal
      </MBModal.SectionHeader>
    )).toMatchSnapshot();
  });

  it('renders with the left and right props specified', () => {
    expect(shallow(
      <MBModal.SectionHeader
        renderLeft={() => <MBModal.Back onClick={() => console.info('back!')} />}
        renderRight={() => <MBModal.Close onClick={() => console.info('close!')} />} >
        Look at this modal
      </MBModal.SectionHeader>
    )).toMatchSnapshot();
  });
});

describe('MBModal.Back', () => {
  it('renders', () => {
    expect(shallow(
      <MBModal.Back onClick={() => console.info('back!')} />
    )).toMatchSnapshot();
  });
});

describe('MBModal.Close', () => {
  it('renders', () => {
    expect(shallow(
      <MBModal.Close onClick={() => console.info('close!')} />
    )).toMatchSnapshot();
  });
});
