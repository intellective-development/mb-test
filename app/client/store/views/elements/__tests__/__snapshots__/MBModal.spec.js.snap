// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MBModal.Back renders 1`] = `
<div
  aria-label="Back"
  className="elMBModal_Back"
  onClick={[Function]}
>
  <MBIcon
    name="back"
  />
  <MBText(span)
    className="elMBModal_BackText"
  >
    Â Back
  </MBText(span)>
</div>
`;

exports[`MBModal.Close renders 1`] = `
<div
  aria-label="Close"
  className="elMBModal_Close"
  onClick={[Function]}
>
  <MBIcon
    name="close"
  />
</div>
`;

exports[`MBModal.Modal renders 1`] = `
<Modal
  autoFocus={true}
  backdrop={true}
  backdropClassName="elMBModal_Backdrop"
  enforceFocus={true}
  keyboard={true}
  manager={
    ModalManager {
      "add": [Function],
      "containers": Array [],
      "data": Array [],
      "handleContainerOverflow": true,
      "hideSiblingNodes": true,
      "isTopModal": [Function],
      "modals": Array [],
      "remove": [Function],
    }
  }
  onHide={[Function]}
  renderBackdrop={[Function]}
  restoreFocus={true}
  show={false}
>
  <div
    className="elMBModal_Wrapper"
  >
    <div
      className="elMBModal_Content elMBModal_Content__Large"
      onClick={[Function]}
    >
      <div>
        My Content!
      </div>
    </div>
  </div>
</Modal>
`;

exports[`MBModal.Modal renders with a header in its content 1`] = `
<Modal
  autoFocus={true}
  backdrop={true}
  backdropClassName="elMBModal_Backdrop"
  enforceFocus={true}
  keyboard={true}
  manager={
    ModalManager {
      "add": [Function],
      "containers": Array [],
      "data": Array [],
      "handleContainerOverflow": true,
      "hideSiblingNodes": true,
      "isTopModal": [Function],
      "modals": Array [],
      "remove": [Function],
    }
  }
  onHide={[Function]}
  renderBackdrop={[Function]}
  restoreFocus={true}
  show={false}
>
  <div
    className="elMBModal_Wrapper"
  >
    <div
      className="elMBModal_Content elMBModal_Content__Large"
      onClick={[Function]}
    >
      <SectionHeader>
        Look at this modal
      </SectionHeader>
      <div>
        My Content!
      </div>
    </div>
  </div>
</Modal>
`;

exports[`MBModal.SectionHeader renders 1`] = `
<div
  className="elMBModal_SectionHeader elMBModal_SectionHeader_Top"
>
  <div
    className="elMBModal_SectionHeader_Filler"
  />
  <MBText(h4)
    className="elMBModal_SectionHeaderText"
  >
    Look at this modal
  </MBText(h4)>
  <div
    className="elMBModal_SectionHeader_Filler"
  />
</div>
`;

exports[`MBModal.SectionHeader renders with the left and right props specified 1`] = `
<div
  className="elMBModal_SectionHeader elMBModal_SectionHeader_Top"
>
  <Back
    onClick={[Function]}
  />
  <MBText(h4)
    className="elMBModal_SectionHeaderText"
  >
    Look at this modal
  </MBText(h4)>
  <Close
    onClick={[Function]}
  />
</div>
`;
